cluster-autoscaler/core/autoscaling_context.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider/builder"
+	"k8s.io/contrib/cluster-autoscaler/expander"
+	"k8s.io/contrib/cluster-autoscaler/simulator"
+
+	kube_client "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	kube_record "k8s.io/kubernetes/pkg/client/record"
+
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	"k8s.io/contrib/cluster-autoscaler/clusterstate"
+	"k8s.io/contrib/cluster-autoscaler/expander/factory"
+	"time"
+)
+
+// AutoscalingContext contains user-configurable constant and configuration-related objects passed to
+// scale up/scale down functions.
+type AutoscalingContext struct {
+	// Options to customize how autoscaling works
+	AutoscalingOptions
+	// CloudProvider used in CA.
+	CloudProvider cloudprovider.CloudProvider
+	// CloudProviderBuilder is used to new CloudProvider
+	CloudProviderBuilder builder.CloudProviderBuilder
+	// ClientSet interface.
+	ClientSet kube_client.Interface
+	// ClusterState for maintaining the state of custer nodes.
+	ClusterStateRegistry *clusterstate.ClusterStateRegistry
+	// Recorder for recording events.
+	Recorder kube_record.EventRecorder
+	// PredicateChecker to check if a pod can fit into a node.
+	PredicateChecker *simulator.PredicateChecker
+	// ExpanderStrategy is the strategy used to choose which node group to expand when scaling up
+	ExpanderStrategy expander.Strategy
+}
+
+// AutoscalingOptions contain various options to customize how autoscaling works
+type AutoscalingOptions struct {
+	// MaxEmptyBulkDelete is a number of empty nodes that can be removed at the same time.
```

Comment:
Thanks for your comment!
I'll take a look again but could you share me if you already caught any non-syncing parts?

I'd been already resolving conflicts periodically while also surveying recent commit diffs each time after a rebase to keep this PR in sync with master.